<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>MolTK C++ API: Class List</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />

<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  <td id="projectlogo"><img alt="Logo" src="moltk55.png"/></td>
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">MolTK C++ API
   &#160;<span id="projectnumber">0.4.0</span>
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.5.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li class="current"><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="inherits.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<div class="title">Class List</div>  </div>
</div>
<div class="contents">
<div class="textblock">Here are the classes, structs, unions and interfaces with brief descriptions:</div><table>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Aligner.html">moltk::Aligner</a></td><td class="indexvalue">Creates macromolecule sequence alignments and structure alignments </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Alignment.html">moltk::Alignment</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1Alignment.html" title="Alignment represents a set of aligned macromolecule sequences and/or structures.">Alignment</a> represents a set of aligned macromolecule sequences and/or structures </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1PDBStructure_1_1Atom.html">moltk::PDBStructure::Atom</a></td><td class="indexvalue">Structure::Atom represents a chemical atom in a molecular structure </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1BaseBiosequence.html">moltk::BaseBiosequence</a></td><td class="indexvalue">Parent class for macromolecule sequences and structures </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1BaseBiosequence_1_1BaseResidue.html">moltk::BaseBiosequence::BaseResidue</a></td><td class="indexvalue">Parent class for macromolecule sequence residues (nucleotides or amino acids) </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1BaseVector3D.html">moltk::BaseVector3D&lt; ELT &gt;</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1BaseVector3D.html" title="BaseVector3D is a parent class for points and vectors.">BaseVector3D</a> is a parent class for points and vectors </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Biosequence.html">moltk::Biosequence</a></td><td class="indexvalue">A macromolecule sequence (DNA or RNA or protein) </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1BiosequenceResidue.html">moltk::BiosequenceResidue</a></td><td class="indexvalue">One residue in a macromolecule sequence (a nucleotide or amino acid) </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Aligner_1_1Cell.html">moltk::Aligner::Cell</a></td><td class="indexvalue">An <a class="el" href="classmoltk_1_1Aligner_1_1Cell.html" title="An Aligner::Cell is one node in the dynamic programming table.">Aligner::Cell</a> is one node in the dynamic programming table </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1PDBStructure_1_1Chain.html">moltk::PDBStructure::Chain</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1PDBStructure_1_1Chain.html" title="Chain represents a single polymer or molecule in a PDBStructure.">Chain</a> represents a single polymer or molecule in a <a class="el" href="classmoltk_1_1PDBStructure.html" title="PDBStructure represents a macromolecule structure as found in a Protein Data Bank file...">PDBStructure</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1EString_1_1const__iterator.html">moltk::EString::const_iterator</a></td><td class="indexvalue">Iterator for accessing the residue positions encoded in an estring </td></tr>
  <tr><td class="indexkey"><a class="el" href="structmoltk_1_1units_1_1Dimension.html">moltk::units::Dimension&lt; M, L, T, Q, N, A, I &gt;</a></td><td class="indexvalue"><a class="el" href="structmoltk_1_1units_1_1Dimension.html" title="Dimension represents a physical dimension such as mass, length, time, etc.">Dimension</a> represents a physical dimension such as mass, length, time, etc </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1EString.html">moltk::EString</a></td><td class="indexvalue">A compact representation of the gapping pattern for one sequence in an alignment </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1MatrixScorer.html">moltk::MatrixScorer</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1MatrixScorer.html" title="MatrixScorer scores alignments using a residue type matrix such as BLOSUM62 or PAM250.">MatrixScorer</a> scores alignments using a residue type matrix such as BLOSUM62 or PAM250 </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1PDBStructure.html">moltk::PDBStructure</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1PDBStructure.html" title="PDBStructure represents a macromolecule structure as found in a Protein Data Bank file...">PDBStructure</a> represents a macromolecule structure as found in a Protein Data Bank file </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Printable.html">moltk::Printable</a></td><td class="indexvalue">Moltk::Printable is an abstract base class to help streamline string conversion in the moltk python bindings </td></tr>
  <tr><td class="indexkey"><a class="el" href="structmoltk_1_1units_1_1Quantity.html">moltk::units::Quantity&lt; U, Y &gt;</a></td><td class="indexvalue"><a class="el" href="structmoltk_1_1units_1_1Quantity.html" title="Quantity represents a physical value with a unit, such as &quot;5.6 nanometers&quot;.">Quantity</a> represents a physical value with a unit, such as "5.6 nanometers" </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Quaternion.html">moltk::Quaternion</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1Quaternion.html" title="Quaternion is an elegant 4-element representation of a 3D rotation.">Quaternion</a> is an elegant 4-element representation of a 3D rotation </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1MatrixScorer_1_1QueryPosition.html">moltk::MatrixScorer::QueryPosition</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1MatrixScorer_1_1QueryPosition.html" title="QueryPosition represents an alignment/sequence column in the second of two sequences being scored by ...">QueryPosition</a> represents an alignment/sequence column in the second of two sequences being scored by a <a class="el" href="classmoltk_1_1MatrixScorer.html" title="MatrixScorer scores alignments using a residue type matrix such as BLOSUM62 or PAM250.">MatrixScorer</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Aligner_1_1QueryPosition.html">moltk::Aligner::QueryPosition</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1Aligner_1_1QueryPosition.html" title="QueryPosition is an alignment column specialized to efficiently compute an alignment score with a Tar...">QueryPosition</a> is an alignment column specialized to efficiently compute an alignment score with a <a class="el" href="classmoltk_1_1Aligner_1_1TargetPosition.html" title="TargetPosition is an alignment column specialized to efficiently compute an alignment score with a Qu...">TargetPosition</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1PDBStructure_1_1Residue.html">moltk::PDBStructure::Residue</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1PDBStructure_1_1Residue.html" title="Residue represents a single amino acid or nucleotide residue, or a small molecule, in a macromolecule structure.">Residue</a> represents a single amino acid or nucleotide residue, or a small molecule, in a macromolecule structure </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1RigidTransform.html">moltk::RigidTransform</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1RigidTransform.html" title="RigidTransform converts one reference frame to another with rotation and translation.">RigidTransform</a> converts one reference frame to another with rotation and translation </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Rotation3D.html">moltk::Rotation3D</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1Rotation3D.html" title="Rotation3D represents a rotation matrix.">Rotation3D</a> represents a rotation matrix </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Rotation3D_1_1Row.html">moltk::Rotation3D::Row</a></td><td class="indexvalue">One row of three numbers in a <a class="el" href="classmoltk_1_1Rotation3D.html" title="Rotation3D represents a rotation matrix.">Rotation3D</a> rotation matrix </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Alignment_1_1Row.html">moltk::Alignment::Row</a></td><td class="indexvalue">Meta-data for one sequence in an <a class="el" href="classmoltk_1_1Alignment.html" title="Alignment represents a set of aligned macromolecule sequences and/or structures.">Alignment</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Aligner_1_1Scorer.html">moltk::Aligner::Scorer</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1Aligner_1_1Scorer.html" title="Scorer can convert, in O(m+n) time, dumb sequence and structure residues into TargetPositions and Que...">Scorer</a> can convert, in O(m+n) time, dumb sequence and structure residues into TargetPositions and QueryPositions, which can efficiently compute alignment scores </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1MatrixScorer_1_1TargetPosition.html">moltk::MatrixScorer::TargetPosition</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1MatrixScorer_1_1TargetPosition.html" title="TargetPosition represents an alignment/sequence column in the first of two sequences being scored by ...">TargetPosition</a> represents an alignment/sequence column in the first of two sequences being scored by a <a class="el" href="classmoltk_1_1MatrixScorer.html" title="MatrixScorer scores alignments using a residue type matrix such as BLOSUM62 or PAM250.">MatrixScorer</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Aligner_1_1TargetPosition.html">moltk::Aligner::TargetPosition</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1Aligner_1_1TargetPosition.html" title="TargetPosition is an alignment column specialized to efficiently compute an alignment score with a Qu...">TargetPosition</a> is an alignment column specialized to efficiently compute an alignment score with a <a class="el" href="classmoltk_1_1Aligner_1_1QueryPosition.html" title="QueryPosition is an alignment column specialized to efficiently compute an alignment score with a Tar...">QueryPosition</a> </td></tr>
  <tr><td class="indexkey"><a class="el" href="structmoltk_1_1units_1_1Unit.html">moltk::units::Unit&lt; D &gt;</a></td><td class="indexvalue"><a class="el" href="structmoltk_1_1units_1_1Unit.html" title="Unit represents a unit of measure, such as gram, nanometer, or second.">Unit</a> represents a unit of measure, such as gram, nanometer, or second </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1UnitVector3D.html">moltk::UnitVector3D</a></td><td class="indexvalue"><a class="el" href="classmoltk_1_1UnitVector3D.html" title="UnitVector3D represents a direction in 3 space, and, unlike Vector3D_&lt;&gt;, does NOT have Units...">UnitVector3D</a> represents a direction in 3 space, and, unlike Vector3D_&lt;&gt;, does NOT have Units </td></tr>
  <tr><td class="indexkey"><a class="el" href="classmoltk_1_1Vector3D__.html">moltk::Vector3D_&lt; ELT &gt;</a></td><td class="indexvalue">Vector3D_&lt;&gt; is an arbitrary point or vector in 3-space </td></tr>
</table>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>



<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.7.5.1
</small></address>

</body>
</html>
