<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>MolTK C++ API: Class Hierarchy</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />

<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  <td id="projectlogo"><img alt="Logo" src="moltk55.png"/></td>
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">MolTK C++ API
   &#160;<span id="projectnumber">0.4.3</span>
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.5.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li class="current"><a href="inherits.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<div class="title">Class Hierarchy</div>  </div>
</div>
<div class="contents">
<div class="textblock">
<p><a href="hierarchy.html">Go to the textual class hierarchy</a></p>
</div><table border="0" cellspacing="10" cellpadding="0">
<tr><td><img src="inherit_graph_0.png" border="0" alt="" usemap="#moltk_1_1Aligner"/>
<map name="moltk_1_1Aligner" id="moltk_1_1Aligner">
<area shape="rect" id="node1" href="classmoltk_1_1Aligner.html" title="The Aligner class creates macromolecule sequence alignments and structure alignments." alt="" coords="6,5,109,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_1.png" border="0" alt="" usemap="#moltk_1_1Alignment___3_01SCORE__TYPE_01_4"/>
<map name="moltk_1_1Alignment___3_01SCORE__TYPE_01_4" id="moltk_1_1Alignment___3_01SCORE__TYPE_01_4">
<area shape="rect" id="node1" href="classmoltk_1_1Alignment__.html" title="Alignment represents a set of aligned macromolecule sequences and/or structures." alt="" coords="6,5,245,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_2.png" border="0" alt="" usemap="#moltk_1_1Alignment___3_01SCORE__TYPE_01_4_1_1Row"/>
<map name="moltk_1_1Alignment___3_01SCORE__TYPE_01_4_1_1Row" id="moltk_1_1Alignment___3_01SCORE__TYPE_01_4_1_1Row">
<area shape="rect" id="node1" href="classmoltk_1_1Alignment___1_1Row.html" title="Meta&#45;data for one sequence in an Alignment_." alt="" coords="6,5,277,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_3.png" border="0" alt="" usemap="#moltk_1_1BaseBiosequence"/>
<map name="moltk_1_1BaseBiosequence" id="moltk_1_1BaseBiosequence">
<area shape="rect" id="node1" href="classmoltk_1_1BaseBiosequence.html" title="Parent class for macromolecule sequences and structures." alt="" coords="6,31,173,57"/><area shape="rect" id="node3" href="classmoltk_1_1Biosequence.html" title="A macromolecule sequence (DNA or RNA or protein)" alt="" coords="245,5,382,32"/><area shape="rect" id="node5" href="classmoltk_1_1PDBStructure_1_1Chain.html" title="Chain represents a single polymer or molecule in a PDBStructure." alt="" coords="222,56,405,83"/></map>
</td></tr>
<tr><td><img src="inherit_graph_4.png" border="0" alt="" usemap="#moltk_1_1BaseBiosequence_1_1BaseResidue"/>
<map name="moltk_1_1BaseBiosequence_1_1BaseResidue" id="moltk_1_1BaseBiosequence_1_1BaseResidue">
<area shape="rect" id="node1" href="classmoltk_1_1BaseBiosequence_1_1BaseResidue.html" title="Parent class for macromolecule sequence residues (nucleotides or amino acids)" alt="" coords="6,31,258,57"/><area shape="rect" id="node3" href="classmoltk_1_1BiosequenceResidue.html" title="One residue in a macromolecule sequence (a nucleotide or amino acid)" alt="" coords="313,5,497,32"/><area shape="rect" id="node5" href="classmoltk_1_1PDBStructure_1_1Residue.html" title="Residue represents a single amino acid or nucleotide residue, or a small molecule, in a macromolecule structure." alt="" coords="307,56,504,83"/></map>
</td></tr>
<tr><td><img src="inherit_graph_5.png" border="0" alt="" usemap="#moltk_1_1BaseVector3D_3_01ELT_01_4"/>
<map name="moltk_1_1BaseVector3D_3_01ELT_01_4" id="moltk_1_1BaseVector3D_3_01ELT_01_4">
<area shape="rect" id="node1" href="classmoltk_1_1BaseVector3D.html" title="BaseVector3D is a parent class for points and vectors." alt="" coords="6,5,199,32"/><area shape="rect" id="node3" href="classmoltk_1_1Vector3D__.html" title="Vector3D_&lt;&gt; is an arbitrary point or vector in 3&#45;space." alt="" coords="248,5,419,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_6.png" border="0" alt="" usemap="#moltk_1_1BaseVector3D_3_01moltk_1_1Real_01_4"/>
<map name="moltk_1_1BaseVector3D_3_01moltk_1_1Real_01_4" id="moltk_1_1BaseVector3D_3_01moltk_1_1Real_01_4">
<area shape="rect" id="node1" href="classmoltk_1_1BaseVector3D.html" title="moltk::BaseVector3D\&lt; moltk::Real \&gt;" alt="" coords="6,5,242,32"/><area shape="rect" id="node3" href="classmoltk_1_1UnitVector3D.html" title="UnitVector3D represents a direction in 3 space, and, unlike Vector3D_&lt;&gt;, does NOT have Units..." alt="" coords="291,5,429,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_7.png" border="0" alt="" usemap="#moltk_1_1dp_1_1AlignmentResult_3_01SCORE__TYPE_01_4"/>
<map name="moltk_1_1dp_1_1AlignmentResult_3_01SCORE__TYPE_01_4" id="moltk_1_1dp_1_1AlignmentResult_3_01SCORE__TYPE_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1AlignmentResult.html" title="Gap structure and score of a pairwise sequence alignment, used as an intermediate value during sequen..." alt="" coords="5,5,296,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_8.png" border="0" alt="" usemap="#moltk_1_1dp_1_1DPCell_3_01SCORE__TYPE_00_01ALIGN__TYPE_00_01GAP__NSEGS_01_4"/>
<map name="moltk_1_1dp_1_1DPCell_3_01SCORE__TYPE_00_01ALIGN__TYPE_00_01GAP__NSEGS_01_4" id="moltk_1_1dp_1_1DPCell_3_01SCORE__TYPE_00_01ALIGN__TYPE_00_01GAP__NSEGS_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1DPCell.html" title="Generic node in dynamic programming table." alt="" coords="5,5,416,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_9.png" border="0" alt="" usemap="#moltk_1_1dp_1_1DPPosition_3_01SCORE__TYPE_00_01dp_1_1DP__ALIGN__GAPPED__ALIGNMENTS_00_01GAP__NSEGS_01_4"/>
<map name="moltk_1_1dp_1_1DPPosition_3_01SCORE__TYPE_00_01dp_1_1DP__ALIGN__GAPPED__ALIGNMENTS_00_01GAP__NSEGS_01_4" id="moltk_1_1dp_1_1DPPosition_3_01SCORE__TYPE_00_01dp_1_1DP__ALIGN__GAPPED__ALIGNMENTS_00_01GAP__NSEGS_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1DPPosition_3_01SCORE__TYPE_00_01dp_1_1DP__ALIGN__GAPPED__ALIGNMENTS_00_01GAP__NSEGS_01_4.html" title="Specialization of alignment column cache for alignment of alignments." alt="" coords="5,5,600,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_10.png" border="0" alt="" usemap="#moltk_1_1dp_1_1DPPosition_3_01SCORE__TYPE_00_01dp_1_1DP__ALIGN__UNGAPPED__SEQUENCES_00_01GAP__NSEGS_01_4"/>
<map name="moltk_1_1dp_1_1DPPosition_3_01SCORE__TYPE_00_01dp_1_1DP__ALIGN__UNGAPPED__SEQUENCES_00_01GAP__NSEGS_01_4" id="moltk_1_1dp_1_1DPPosition_3_01SCORE__TYPE_00_01dp_1_1DP__ALIGN__UNGAPPED__SEQUENCES_00_01GAP__NSEGS_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1DPPosition_3_01SCORE__TYPE_00_01dp_1_1DP__ALIGN__UNGAPPED__SEQUENCES_00_01GAP__NSEGS_01_4.html" title="Specialization of alignment column cache for alignment two individual sequences." alt="" coords="5,5,616,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_11.png" border="0" alt="" usemap="#moltk_1_1dp_1_1DPTable_3_01SCORE__TYPE_00_01DP__MEMORY__LARGE_00_01ALIGN__TYPE_00_011_01_4"/>
<map name="moltk_1_1dp_1_1DPTable_3_01SCORE__TYPE_00_01DP__MEMORY__LARGE_00_01ALIGN__TYPE_00_011_01_4" id="moltk_1_1dp_1_1DPTable_3_01SCORE__TYPE_00_01DP__MEMORY__LARGE_00_01ALIGN__TYPE_00_011_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1DPTable_3_01SCORE__TYPE_00_01DP__MEMORY__LARGE_00_01ALIGN__TYPE_00_011_01_4.html" title="template specialization of DPTable for Affine alignment in large memory." alt="" coords="6,5,495,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_12.png" border="0" alt="" usemap="#moltk_1_1dp_1_1GapScore_3_01SCORE__TYPE_00_011_01_4"/>
<map name="moltk_1_1dp_1_1GapScore_3_01SCORE__TYPE_00_011_01_4" id="moltk_1_1dp_1_1GapScore_3_01SCORE__TYPE_00_011_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1GapScore_3_01SCORE__TYPE_00_011_01_4.html" title="template specialization for affine gap score function" alt="" coords="6,5,274,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_13.png" border="0" alt="" usemap="#moltk_1_1dp_1_1GapScore_3_01SCORE__TYPE_00_01GAP__NSEGS_01_4"/>
<map name="moltk_1_1dp_1_1GapScore_3_01SCORE__TYPE_00_01GAP__NSEGS_01_4" id="moltk_1_1dp_1_1GapScore_3_01SCORE__TYPE_00_01GAP__NSEGS_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1GapScore.html" title="Parameters for scoring indel gaps in sequence alignments." alt="" coords="6,5,349,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_14.png" border="0" alt="" usemap="#moltk_1_1dp_1_1RunningGapScore_3_01SCORE__TYPE_00_01DP__ALIGN__GAPPED__ALIGNMENTS_00_011_01_4"/>
<map name="moltk_1_1dp_1_1RunningGapScore_3_01SCORE__TYPE_00_01DP__ALIGN__GAPPED__ALIGNMENTS_00_011_01_4" id="moltk_1_1dp_1_1RunningGapScore_3_01SCORE__TYPE_00_01DP__ALIGN__GAPPED__ALIGNMENTS_00_011_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1RunningGapScore_3_01SCORE__TYPE_00_01DP__ALIGN__GAPPED__ALIGNMENTS_00_011_01_4.html" title="Template specialization for alignment of alignments affine gap running score (E or F in Gusfield recu..." alt="" coords="5,5,544,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_15.png" border="0" alt="" usemap="#moltk_1_1dp_1_1RunningGapScore_3_01SCORE__TYPE_00_01DP__ALIGN__UNGAPPED__SEQUENCES_00_011_01_4"/>
<map name="moltk_1_1dp_1_1RunningGapScore_3_01SCORE__TYPE_00_01DP__ALIGN__UNGAPPED__SEQUENCES_00_011_01_4" id="moltk_1_1dp_1_1RunningGapScore_3_01SCORE__TYPE_00_01DP__ALIGN__UNGAPPED__SEQUENCES_00_011_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1RunningGapScore_3_01SCORE__TYPE_00_01DP__ALIGN__UNGAPPED__SEQUENCES_00_011_01_4.html" title="Template specialization for single sequence affine gap running score (E or F in Gusfield recurrence)..." alt="" coords="5,5,560,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_16.png" border="0" alt="" usemap="#moltk_1_1dp_1_1RunningScore_3_01SCORE__TYPE_01_4"/>
<map name="moltk_1_1dp_1_1RunningScore_3_01SCORE__TYPE_01_4" id="moltk_1_1dp_1_1RunningScore_3_01SCORE__TYPE_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1dp_1_1RunningScore.html" title="Generic accumulated score in a DPCell." alt="" coords="6,5,282,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_17.png" border="0" alt="" usemap="#moltk_1_1EString"/>
<map name="moltk_1_1EString" id="moltk_1_1EString">
<area shape="rect" id="node1" href="classmoltk_1_1EString.html" title="A compact representation of the gapping pattern for one sequence in an alignment." alt="" coords="6,5,111,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_18.png" border="0" alt="" usemap="#moltk_1_1EString_1_1const__iterator"/>
<map name="moltk_1_1EString_1_1const__iterator" id="moltk_1_1EString_1_1const__iterator">
<area shape="rect" id="node1" href="classmoltk_1_1EString_1_1const__iterator.html" title="Iterator for accessing the residue positions encoded in an estring." alt="" coords="5,5,197,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_19.png" border="0" alt="" usemap="#moltk_1_1MatrixScorer___3_01SCORE__TYPE_00_01GAP__NSEGS_01_4"/>
<map name="moltk_1_1MatrixScorer___3_01SCORE__TYPE_00_01GAP__NSEGS_01_4" id="moltk_1_1MatrixScorer___3_01SCORE__TYPE_00_01GAP__NSEGS_01_4">
<area shape="rect" id="node1" href="classmoltk_1_1MatrixScorer__.html" title="MatrixScorer scores alignments using a residue type matrix such as BLOSUM62 or PAM250." alt="" coords="6,5,349,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_20.png" border="0" alt="" usemap="#moltk_1_1PDBStructure"/>
<map name="moltk_1_1PDBStructure" id="moltk_1_1PDBStructure">
<area shape="rect" id="node1" href="classmoltk_1_1PDBStructure.html" title="PDBStructure represents a macromolecule structure as found in a Protein Data Bank file..." alt="" coords="6,5,149,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_21.png" border="0" alt="" usemap="#moltk_1_1PDBStructure_1_1Atom"/>
<map name="moltk_1_1PDBStructure_1_1Atom" id="moltk_1_1PDBStructure_1_1Atom">
<area shape="rect" id="node1" href="classmoltk_1_1PDBStructure_1_1Atom.html" title="Structure::Atom represents a chemical atom in a molecular structure." alt="" coords="5,5,187,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_22.png" border="0" alt="" usemap="#moltk_1_1Quaternion"/>
<map name="moltk_1_1Quaternion" id="moltk_1_1Quaternion">
<area shape="rect" id="node1" href="classmoltk_1_1Quaternion.html" title="Quaternion is an elegant 4&#45;element representation of a 3D rotation." alt="" coords="5,5,131,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_23.png" border="0" alt="" usemap="#moltk_1_1RigidTransform"/>
<map name="moltk_1_1RigidTransform" id="moltk_1_1RigidTransform">
<area shape="rect" id="node1" href="classmoltk_1_1RigidTransform.html" title="RigidTransform converts one reference frame to another with rotation and translation." alt="" coords="6,5,154,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_24.png" border="0" alt="" usemap="#moltk_1_1Rotation3D"/>
<map name="moltk_1_1Rotation3D" id="moltk_1_1Rotation3D">
<area shape="rect" id="node1" href="classmoltk_1_1Rotation3D.html" title="Rotation3D represents a rotation matrix." alt="" coords="6,5,133,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_25.png" border="0" alt="" usemap="#moltk_1_1Rotation3D_1_1Row"/>
<map name="moltk_1_1Rotation3D_1_1Row" id="moltk_1_1Rotation3D_1_1Row">
<area shape="rect" id="node1" href="classmoltk_1_1Rotation3D_1_1Row.html" title="One row of three numbers in a Rotation3D rotation matrix." alt="" coords="6,5,165,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_26.png" border="0" alt="" usemap="#moltk_1_1SubstitutionMatrix___3_01SCORE__TYPE_01_4"/>
<map name="moltk_1_1SubstitutionMatrix___3_01SCORE__TYPE_01_4" id="moltk_1_1SubstitutionMatrix___3_01SCORE__TYPE_01_4">
<area shape="rect" id="node1" href="classmoltk_1_1SubstitutionMatrix__.html" title="SubstitutionMatrix scores alignments using a residue type matrix such as BLOSUM62 or PAM250..." alt="" coords="5,5,291,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_27.png" border="0" alt="" usemap="#moltk_1_1units_1_1Dimension_3_01M_00_01L_00_01T_00_01Q_00_01N_00_01A_00_01I_01_4"/>
<map name="moltk_1_1units_1_1Dimension_3_01M_00_01L_00_01T_00_01Q_00_01N_00_01A_00_01I_01_4" id="moltk_1_1units_1_1Dimension_3_01M_00_01L_00_01T_00_01Q_00_01N_00_01A_00_01I_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1units_1_1Dimension.html" title="Dimension represents a physical dimension such as mass, length, time, etc." alt="" coords="6,5,293,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_28.png" border="0" alt="" usemap="#moltk_1_1units_1_1Quantity_3_01U_00_01Y_01_4"/>
<map name="moltk_1_1units_1_1Quantity_3_01U_00_01Y_01_4" id="moltk_1_1units_1_1Quantity_3_01U_00_01Y_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1units_1_1Quantity.html" title="Quantity represents a physical value with a unit, such as &quot;5.6 nanometers&quot;." alt="" coords="5,5,203,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_29.png" border="0" alt="" usemap="#moltk_1_1units_1_1Unit_3_01D_01_4"/>
<map name="moltk_1_1units_1_1Unit_3_01D_01_4" id="moltk_1_1units_1_1Unit_3_01D_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1units_1_1Unit.html" title="Unit represents a unit of measure, such as gram, nanometer, or second." alt="" coords="5,5,160,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_30.png" border="0" alt="" usemap="#moltk_1_1units_1_1value__finder_3_01Quantity_3_01U_00_01Y_01_4_01_4"/>
<map name="moltk_1_1units_1_1value__finder_3_01Quantity_3_01U_00_01Y_01_4_01_4" id="moltk_1_1units_1_1value__finder_3_01Quantity_3_01U_00_01Y_01_4_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1units_1_1value__finder_3_01Quantity_3_01U_00_01Y_01_4_01_4.html" title="Template specialization for zero and infinit Quantity values." alt="" coords="5,5,293,32"/></map>
</td></tr>
<tr><td><img src="inherit_graph_31.png" border="0" alt="" usemap="#moltk_1_1units_1_1value__finder_3_01T_01_4"/>
<map name="moltk_1_1units_1_1value__finder_3_01T_01_4" id="moltk_1_1units_1_1value__finder_3_01T_01_4">
<area shape="rect" id="node1" href="structmoltk_1_1units_1_1value__finder.html" title="Generalized templated methods to help template classes get zero or infinity values from both moltk::u..." alt="" coords="6,5,202,32"/></map>
</td></tr>
</table>
</div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>



<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.7.5.1
</small></address>

</body>
</html>
